File: 第三阶段/3.7-泛型typing与CSharp泛型/GenericDemo.cs
// C# 泛型示例
// 对应文章：3.7 泛型typing与CSharp泛型

using System;
using System.Collections.Generic;
using System.Linq;

public class GenericDemo
{
    // 1. 泛型类
    public class Stack<T>
    {
        private readonly List<T> _items = new();

        public void Push(T item) => _items.Add(item);

        public T Pop()
        {
            if (_items.Count == 0)
                throw new InvalidOperationException("栈为空");
            var item = _items[^1];
            _items.RemoveAt(_items.Count - 1);
            return item;
        }

        public T Peek() => _items[^1];
        public bool IsEmpty => _items.Count == 0;
        public int Count => _items.Count;
    }

    // 2. 泛型函数
    static T? First<T>(List<T> items) where T : class
        => items.FirstOrDefault();

    static Dictionary<K, V> Merge<K, V>(Dictionary<K, V> d1, Dictionary<K, V> d2)
    {
        var result = new Dictionary<K, V>(d1);
        foreach (var kv in d2) result[kv.Key] = kv.Value;
        return result;
    }

    // 3. 类型约束
    static T Add<T>(T a, T b) where T : INumber<T>
        => a + b;

    // 接口约束示例
    public interface IRepository<T> where T : class
    {
        void Add(T item);
        List<T> GetAll();
        T? Find(Func<T, bool> predicate);
    }

    // 4. 实际场景：泛型仓储
    public record User(int Id, string Name);
    public record Product(int Id, string Title, decimal Price);

    public class Repository<T> : IRepository<T> where T : class
    {
        private readonly List<T> _items = new();

        public void Add(T item) => _items.Add(item);
        public List<T> GetAll() => _items.ToList();
        public T? Find(Func<T, bool> predicate) => _items.FirstOrDefault(predicate);
    }

    static void Main()
    {
        Console.WriteLine("===== 泛型：typing.Generic vs C# 泛型 =====\n");

        // 1. 泛型栈
        Console.WriteLine("--- 1. 泛型栈 ---");
        var intStack = new Stack<int>();
        intStack.Push(1);
        intStack.Push(2);
        Console.WriteLine($"整数栈弹出: {intStack.Pop()}");

        var strStack = new Stack<string>();
        strStack.Push("hello");
        Console.WriteLine($"字符串栈弹出: {strStack.Pop()}");

        // 2. 泛型函数
        Console.WriteLine("\n--- 2. 泛型函数 ---");
        var nums = new List<int> { 1, 2, 3 };
        Console.WriteLine($"first: {First(nums)}");

        var d1 = new Dictionary<string, int> { ["a"] = 1 };
        var d2 = new Dictionary<string, int> { ["b"] = 2 };
        var merged = Merge(d1, d2);
        Console.WriteLine($"merge: {string.Join(", ", merged.Select(kv => $"{kv.Key}:{kv.Value}"))}");

        // 3. 泛型仓储
        Console.WriteLine("\n--- 3. 泛型仓储 ---");
        var userRepo = new Repository<User>();
        userRepo.Add(new User(1, "Alice"));
        userRepo.Add(new User(2, "Bob"));

        var productRepo = new Repository<Product>();
        productRepo.Add(new Product(1, "Laptop", 999.99m));

        var user = userRepo.Find(u => u.Name == "Alice");
        Console.WriteLine($"用户: {user}");

        var product = productRepo.Find(p => p.Price > 500);
        Console.WriteLine($"商品: {product}");
    }
}

File: 第三阶段/3.7-泛型typing与CSharp泛型/generic_demo.py
# Python 泛型 typing 示例
# 对应文章：3.7 泛型typing与CSharp泛型

from typing import (
    TypeVar, Generic, List, Dict, Optional,
    Callable, Protocol, Type, Any
)
from dataclasses import dataclass

# ========== 1. TypeVar —— 类型变量 ==========
T = TypeVar("T")
K = TypeVar("K")
V = TypeVar("V")

# ========== 2. 泛型类 ==========
class Stack(Generic[T]):
    """泛型栈"""
    def __init__(self) -> None:
        self._items: List[T] = []

    def push(self, item: T) -> None:
        self._items.append(item)

    def pop(self) -> T:
        if not self._items:
            raise IndexError("栈为空")
        return self._items.pop()

    def peek(self) -> T:
        return self._items[-1]

    def is_empty(self) -> bool:
        return len(self._items) == 0

    def __len__(self) -> int:
        return len(self._items)

# 使用
int_stack: Stack[int] = Stack()
int_stack.push(1)
int_stack.push(2)
print(f"整数栈弹出: {int_stack.pop()}")  # 2

str_stack: Stack[str] = Stack()
str_stack.push("hello")
print(f"字符串栈弹出: {str_stack.pop()}")  # hello

# ========== 3. 泛型函数 ==========
def first(items: List[T]) -> Optional[T]:
    """返回列表第一个元素"""
    return items[0] if items else None

def merge(dict1: Dict[K, V], dict2: Dict[K, V]) -> Dict[K, V]:
    """合并两个字典"""
    return {**dict1, **dict2}

print(f"first: {first([1, 2, 3])}")
print(f"merge: {merge({'a': 1}, {'b': 2})}")

# ========== 4. 类型约束 (bound) ==========
Number = TypeVar("Number", int, float, complex)

def add(a: Number, b: Number) -> Number:
    return a + b

print(f"add(1, 2) = {add(1, 2)}")
print(f"add(1.5, 2.5) = {add(1.5, 2.5)}")

# ========== 5. Protocol —— 结构化子类型（鸭子类型的类型化） ==========
class Drawable(Protocol):
    def draw(self) -> str: ...

class Circle:
    def draw(self) -> str:
        return "○"

class Square:
    def draw(self) -> str:
        return "□"

def draw_shape(shape: Drawable) -> None:
    print(f"绘制: {shape.draw()}")

draw_shape(Circle())  # [OK] Circle 有 draw 方法
draw_shape(Square())  # [OK] Square 有 draw 方法
# draw_shape("hello")  # [X] str 没有 draw 方法

# ========== 6. 实际场景：泛型仓储模式 ==========
@dataclass
class User:
    id: int
    name: str

@dataclass
class Product:
    id: int
    title: str
    price: float

class Repository(Generic[T]):
    def __init__(self, items: Optional[List[T]] = None):
        self._items = items or []

    def add(self, item: T) -> None:
        self._items.append(item)

    def get_all(self) -> List[T]:
        return self._items.copy()

    def find(self, predicate: Callable[[T], bool]) -> Optional[T]:
        return next((item for item in self._items if predicate(item)), None)

user_repo: Repository[User] = Repository()
user_repo.add(User(1, "Alice"))
user_repo.add(User(2, "Bob"))

product_repo: Repository[Product] = Repository()
product_repo.add(Product(1, "Laptop", 999.99))

print(f"用户: {user_repo.find(lambda u: u.name == 'Alice')}")
print(f"商品: {product_repo.find(lambda p: p.price > 500)}")
